File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        longest = 0
        left = 0
        
        for right in range(0, len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1
                
            char_set.add(s[right])
            longest = max(longest, len(char_set))
                          
        return longest

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {

    const nonRepeating = new Set();
    let ans = 0;
    let left = 0;
    
    for (let right in s) {
        // console.log(right)
        while (nonRepeating.has(s[right])) {
            nonRepeating.delete(s[left]);
            left += 1;
            // console.log(nonRepeating)
        }
        
        nonRepeating.add(s[right]);
        
        // console.log(nonRepeating, "after")
        
        ans = Math.max(ans, right - left + 1)
        // console.log("ans", ans)
    }
    return ans;
};